### ISSUE TITLE
[Swagger] @id gets generated as int, but exported as string

### ISSUE THREAD

**Issue by dkarlovi** [api-platform/core#758]
I'm using 2.0.0-RC1 and have an entity called `Ingredient` with declaration as so

```
/**
 * Ingredient.
 *
 * @ORM\Table(name="ingredient")
 * @ORM\Entity(repositoryClass="AppBundle\Repository\IngredientRepository")
 * @ApiResource(iri="http://schema.org/Ingredient")
 */
class Ingredient
{
    /**
     * @var int
     *
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;
}
```

Swagger definition in `apidoc.json` says

```
definitions: {
    Ingredient: {
        type: "object",
        xml: {
            name: "response"
        },
        properties: {
            id: {
                type: "integer"
            },
```

so it expects it to be an `int`. But, when accessing the API, ID is set as:

```
"@id": "/api/ingredients/4",
```

which makes it a string. This means that using [swagger-codegen](https://github.com/swagger-api/swagger-codegen) creates the API client which will cast all ID's to 0, as it's using (in PHP language mode)

```
// $data = /api/ingredients/4
// $class = int
settype($data, $class);
```

which reproduces as:

```
$ php70 -r '$s = "/api/ingredients/4"; settype($s, "int"); var_dump($s);'
int(0)
```

**Comment by dunglas** [comment]
`@id` and `id` aren't the same property. Sometimes you need you need to access (often in write context) the raw id (e.g. it's a you want to set, or an ISBN number, or an existing, non-autogenerated id). In such case, the raw `id` need to be exposed through the API

The JSON-LD `@id` tag is different. It's a special tag used only in JSON-LD containing the document's IRI. It is not documented in Swagger (because Swagger doesn't allow to have a varying model definition by Content-Type).

**Comment by dkarlovi** [comment]
@dunglas I see, thanks. If it's JSON+LD only, it's a bug as I'm getting:

    "id": "/api/ingredients/4",

with `Accept: application/json` too.

**Comment by dunglas** [comment]
Humm indeed we generate IRI for all formats. Not sure if you should continue to that or not. Anyway we need to do something: remove IRI from non-hypermedia formats or change the type in Swagger. Any opinion @api-platform/core-team?

**Comment by damijank** [comment]
It's even weirder than that. Now I've exposed the raw `id` through the API.
The JSON+LD correctly gives me `@id` (as IRI) **and** `id` (as integer or whatever type the `id` is).
The JSON response now gives me the `id` as integer, too. Prior to exposing, as @dkarlovi said, the `id` in plain JSON was an IRI (string) but documented as integer.

**Comment by teohhanhui** [comment]
There should not be an `id` property in the response if it's not being exposed (via serializer groups). `@id` should be present for JSON-LD but not for plain JSON, I don't know enough about the other formats to comment.

**Comment by teohhanhui** [comment]
Having the IRI in `id` property is just wrong.

**Comment by Simperfit** [comment]
This has been fixed by #772 @dkarlovi can you confirm ?
